common-libs/common/src/shared/DidResolver.ts

```diff
@@ -0,0 +1,57 @@
+import { RegistryApiService } from '@affinidi/internal-api-clients'
+
+type CachedPromises = Map<string, Promise<Record<string, any>>>
+type ServiceWithCache = {
+  service: RegistryApiService
+  cache: CachedPromises
+}
+
+type ConstructorOptions = ConstructorParameters<typeof RegistryApiService>[0]
+
+const services: Map<string, ServiceWithCache> = new Map()
+
+const getService = (options: ConstructorOptions) => {
+  const { accessApiKey, registryUrl } = options
+  const cacheKey = JSON.stringify({ accessApiKey, registryUrl })
+  if (!services.has(cacheKey)) {
+    services.set(cacheKey, {
+      service: new RegistryApiService(options),
+      cache: new Map(),
+    })
+  }
+
+  return services.get(cacheKey)
+}
+
+const resolveDid = ({ service, cache }: ServiceWithCache, did: string) => {
+  if (!cache.has(did)) {
+    const promise = new Promise((resolve, reject) => {
+      service
+        .resolveDid({ did })
+        .then(({ body }) => {
+          const { didDocument } = body
+          resolve(didDocument)
+        })
+        .catch((reason) => {
+          // Don't cache fails
+          cache.delete(did)
+          reject(reason)
+        })
+    })
+    cache.set(did, promise)
+  }
+
+  return cache.get(did)
+}
+
+export class DidResolver {
+  private readonly _service: ServiceWithCache
+
+  constructor(options: ConstructorOptions) {
+    this._service = getService(options)
+  }
+
+  resolveDid(did: string) {
+    return resolveDid(this._service, did)
+  }
+}
```

Comment:
It feels like it's going to be a memory leak

Reply:
It stores:
* For every combination of serviceUrl + accessToken that was ever passed to it: a service instance and a cached set of promises (initially empty)
* For every service instance and every did that was ever passed to it: a promise with the resolved DID.

If some application that uses SDK runs for a long time and resolves a lot of different DIDs, the cache will grow, obviously (because it will have to store all the DIDs it ever resolved in this process). Is that the leak you're talking about?

Was our SDK optimized for that kind of workload anyway?

Reply:
Yeah, exactly. 

> Was our SDK optimized for that kind of workload anyway?

I don't know, but it seems logical to me to expire some values eventually, or have some limit to the cache, because we don't know how our SDK is used - it might as well be used to resolve a lot of DIDs - or it's an edge case?

Reply:
I think it's an edge case, and implementing an expiration is probably more trouble than would be worth it.

Reply:
Okay, let's start with the current solution and see if there are any real problems.